File: monitor_customhmd/DriverComm/ShMem.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.IO.MemoryMappedFiles;
using System.Runtime.InteropServices;
using System.Security.AccessControl;
using System.Security.Principal;
using System.Threading;

namespace monitor_customhmd.DriverComm
{
    [Flags]
    public enum CommState
    {
        Uninitialized = 0,
        TrackerActive = 1,
        DriverActive = 2
    }

    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi, Pack = 1)]
    public struct CommStatus
    {
        public CommState State;
        public uint DriverTime;
        public int EnableWatchDog;
        public int IncomingPackets;
        public int OutgoingPackets;
    };

    public class ShMem : IDisposable
    {
        //private static int _screenBufferSize = Marshal.SizeOf(typeof(PacketInfo)) + (3840 * 2160 * 4); //ScreenInfo + 4K display
        private static int _commBufferSize = 1024 * 1024;
        private static int _statusSize = Marshal.SizeOf(typeof(CommStatus));
        private static int _packetSize = 32;
        private static int _maxPackets = 16;
        private static int _statusOffset = 0;
        private static int _incomingOffset = _statusSize;
        private static int _outgoingOffset = _incomingOffset + (_packetSize * _maxPackets);
        private uint _driverTimestamp = 0;
        private uint _prevDriverTimestamp = 0;

        private DateTime DriverTime = DateTime.MinValue;

        private CommStatus _status;
        public CommStatus Status { get { return _status; } }



        private Mutex _commAccessLock;
        private MemoryMappedFile _commSharedMem;
        private MemoryMappedViewAccessor _commAccessor;

        //private Mutex _screenAccessLock;        
        //private MemoryMappedFile _screenSharedMem;         
        //private MemoryMappedViewAccessor _screenAccessor;

        public ShMem()
        {
            var newMutex = false;
            var mutexSecurity = new MutexSecurity(); mutexSecurity.AddAccessRule(new MutexAccessRule(new SecurityIdentifier(WellKnownSidType.WorldSid, null), MutexRights.FullControl, AccessControlType.Allow));
            var memSecurity = new MemoryMappedFileSecurity(); memSecurity.AddAccessRule(new AccessRule<MemoryMappedFileRights>(new SecurityIdentifier(WellKnownSidType.WorldSid, null), MemoryMappedFileRights.FullControl, AccessControlType.Allow));

            _commAccessLock = new Mutex(false, "Global\\CustomHMDCommLock", out newMutex, mutexSecurity);
            _commSharedMem = MemoryMappedFile.CreateOrOpen("CustomHMDComm", _commBufferSize, MemoryMappedFileAccess.ReadWrite, MemoryMappedFileOptions.None, memSecurity, System.IO.HandleInheritability.Inheritable);
            _commAccessor = _commSharedMem.CreateViewAccessor();

            //_screenAccessLock = new Mutex(false, "Global\\CustomHMDScreenLock", out newMutex, mutexSecurity);
            //_screenSharedMem = MemoryMappedFile.CreateOrOpen("CustomHMDLeft", _screenBufferSize, MemoryMappedFileAccess.ReadWrite, MemoryMappedFileOptions.None, memSecurity, System.IO.HandleInheritability.Inheritable);
            //_screenAccessor = _screenSharedMem.CreateViewAccessor();

            //read last status from shmem
            if (_commAccessLock.WaitOne(1000))
            {
                _commAccessor.Read(_statusOffset, out _status);
                _commAccessLock.ReleaseMutex();
            }

        }

        public void Dispose()
        {
            //UpdateState(CommState.TrackerConnected);

            _commAccessor.Dispose();
            _commAccessor = null;
            _commSharedMem.Dispose();
            _commSharedMem = null;
            _commAccessLock.Dispose();
            _commAccessLock = null;

            //_screenAccessor.Dispose();
            //_screenAccessor = null;
            //_screenSharedMem.Dispose();
            //_screenSharedMem = null;
            //_screenAccessLock.Dispose();
            //_screenAccessLock = null;
        }

        public CommState State
        {
            get
            {
                return _status.State;
            }
            set
            {
                if (_commAccessLock.WaitOne(100))
                {
                    _commAccessor.Read(_statusOffset, out _status);
                    _status.State = value;
                    _driverTimestamp = _status.DriverTime;
                    if (_driverTimestamp != _prevDriverTimestamp) DriverTime = DateTime.Now;
                    _prevDriverTimestamp = _driverTimestamp;
                    _commAccessor.Write(_statusOffset, ref _status);
                    _commAccessLock.ReleaseMutex();
                }
            }
        }

        public bool IsDriverActive
        {
            get
            {
                return (DateTime.Now - DriverTime).TotalSeconds < 2;
            }
        }

        public void EnableWatchDog(bool en)
        {
            if (_commAccessLock.WaitOne(100))
            {
                _commAccessor.Read(_statusOffset, out _status);
                _driverTimestamp = _status.DriverTime;
                if (_driverTimestamp != _prevDriverTimestamp) DriverTime = DateTime.Now;
                _prevDriverTimestamp = _driverTimestamp;
                _status.EnableWatchDog = en ? 1 : 0;
                _commAccessor.Write(_statusOffset, ref _status);
                _commAccessLock.ReleaseMutex();
            }
        }

        public void WriteIncomingPacket(byte[] packet, int dataOffset = 0)
        {
            if (_commAccessLock.WaitOne(100))
            {
                _commAccessor.Read(_statusOffset, out _status);
                _driverTimestamp = _status.DriverTime;
                if (_driverTimestamp != _prevDriverTimestamp) DriverTime = DateTime.Now;
                _prevDriverTimestamp = _driverTimestamp;
                if (_status.IncomingPackets < _maxPackets)
                {
                    var offset = _incomingOffset + (_status.IncomingPackets * _packetSize);
                    for (int i = 0; i < _packetSize; i++)
                        _commAccessor.Write(offset + i, ref packet[i + dataOffset]);
                    _status.IncomingPackets++;
                    _commAccessor.Write(_statusOffset, ref _status);
                }
                else
                {
                    //reset buffer
                    Debug.WriteLine("Buffer full");
                    _status.IncomingPackets = 0;
                    _commAccessor.Write(_statusOffset, ref _status);
                }
                _commAccessLock.ReleaseMutex();
            }
        }

        public List<byte[]> ReadOutgoingPackets(int dataOffset = 0)
        {
            List<byte[]> result = null;
            if (_commAccessLock.WaitOne(100))
            {
                _commAccessor.Read(_statusOffset, out _status);
                _driverTimestamp = _status.DriverTime;
                if (_driverTimestamp != _prevDriverTimestamp) DriverTime = DateTime.Now;
                _prevDriverTimestamp = _driverTimestamp;
                if (_status.OutgoingPackets > 0)
                {
                    result = new List<byte[]>();
                    for (var p = 0; p < _status.OutgoingPackets; p++)
                    {
                        byte[] data = new byte[32 + dataOffset];
                        var currOffset = _outgoingOffset + (p * _packetSize);
                        for (var i = 0; i < _packetSize; i++)
                            _commAccessor.Read(currOffset + i, out data[i + dataOffset]);
                        result.Add(data);
                    }
                    _status.OutgoingPackets = 0;
                    _commAccessor.Write(_statusOffset, ref _status);
                }
                _commAccessLock.ReleaseMutex();
            }
            return result;
        }

        //public int GetScreenImage(out byte[] bitmap)
        //{
        //    bitmap = null;

        //    if (!IsDriverActive)
        //        return 0;
        //    var infoSize = Marshal.SizeOf(typeof(PacketInfo));
        //    PacketInfo info;
        //    int size = 0;
        //    if (_screenAccessLock.WaitOne(10))
        //    {
        //        _screenAccessor.Read(0, out info);
        //        size = info.Size;
        //        if (size > 0)
        //        {
        //            using (var stream = _screenSharedMem.CreateViewStream(infoSize, size))
        //            {
        //                using (BinaryReader binReader = new BinaryReader(stream))
        //                {
        //                    bitmap = binReader.ReadBytes(size);
        //                }
        //            }
        //        }
        //        info.Size = 0;
        //        _screenAccessor.Write(0, ref info);
        //        _screenAccessLock.ReleaseMutex();
        //    }
        //    return size;
        //}
    }
}

File: driver_customhmd/ShMem.cpp
#include "ShMem.h"
#include <aclapi.h>

CShMem::CShMem()
{
	WatchDogEnabled = false;
	_commAccessLock = CreateMutex(nullptr, false, L"Global\\CustomHMDCommLock");
	_commAccessor = nullptr;
	_commSharedMem = CreateFileMapping(INVALID_HANDLE_VALUE, nullptr, PAGE_READWRITE, 0, _commBufferSize, L"CustomHMDComm");
	if (_commSharedMem)
	{
		SetSecurityInfo(_commSharedMem, SE_KERNEL_OBJECT, DACL_SECURITY_INFORMATION, 0, 0, (PACL)nullptr, nullptr);
		_commAccessor = (char *)MapViewOfFile(_commSharedMem, FILE_MAP_ALL_ACCESS, 0, 0, _commBufferSize);
	}

	_screenAccessLock[0] = CreateMutex(nullptr, false, L"Global\\CustomHMDLeftLock");
	_screenAccessor[0] = nullptr;
	_screenSharedMem[0] = CreateFileMapping(INVALID_HANDLE_VALUE, nullptr, PAGE_READWRITE, 0, _screenBufferSize, L"CustomHMDLeft");
	if (_screenSharedMem[0])
	{
		SetSecurityInfo(_screenSharedMem[0], SE_KERNEL_OBJECT, DACL_SECURITY_INFORMATION, 0, 0, (PACL)nullptr, nullptr);
		_screenAccessor[0] = (char *)MapViewOfFile(_screenSharedMem[0], FILE_MAP_ALL_ACCESS, 0, 0, _screenBufferSize);
	}

	_screenAccessLock[1] = CreateMutex(nullptr, false, L"Global\\CustomHMDRightLock");
	_screenAccessor[1] = nullptr;
	_screenSharedMem[1] = CreateFileMapping(INVALID_HANDLE_VALUE, nullptr, PAGE_READWRITE, 0, _screenBufferSize, L"CustomHMDRight");
	if (_screenSharedMem[1])
	{
		SetSecurityInfo(_screenSharedMem[1], SE_KERNEL_OBJECT, DACL_SECURITY_INFORMATION, 0, 0, (PACL)nullptr, nullptr);
		_screenAccessor[1] = (char *)MapViewOfFile(_screenSharedMem[1], FILE_MAP_ALL_ACCESS, 0, 0, _screenBufferSize);
	}
}


CShMem::~CShMem()
{
	if (_commAccessor) UnmapViewOfFile(_commAccessor);
	_commAccessor = nullptr;
	if (_commSharedMem) CloseHandle(_commSharedMem);
	_commSharedMem = nullptr;
	CloseHandle(_commAccessLock);
	_commAccessLock = nullptr;

	if (_screenAccessor[0]) UnmapViewOfFile(_screenAccessor[0]);
	if (_screenAccessor[1]) UnmapViewOfFile(_screenAccessor[1]);
	_screenAccessor[0] = nullptr; _screenAccessor[1] = nullptr;
	if (_screenSharedMem[0]) CloseHandle(_screenSharedMem[0]);
	if (_screenSharedMem[1]) CloseHandle(_screenSharedMem[1]);
	_screenSharedMem[0] = nullptr; _screenSharedMem[1] = nullptr;
	CloseHandle(_screenAccessLock[0]); CloseHandle(_screenAccessLock[1]);
	_screenAccessLock[0] = nullptr; _screenAccessLock[1] = nullptr;

}

CommState CShMem::GetState()
{
	if (WaitForSingleObject(_commAccessLock, 100) == WAIT_OBJECT_0)
	{
		_status = *((CommStatus*)&_commAccessor[_statusOffset]);
		WatchDogEnabled = _status.EnableWatchDog ? true : false;
		ReleaseMutex(_commAccessLock);
	}	
	return _status.State;		
}

void CShMem::WriteOutgoingPacket(char *packet)
{
	if (WaitForSingleObject(_commAccessLock, 100) == WAIT_OBJECT_0)
	{
		_status = *((CommStatus*)&_commAccessor[_statusOffset]);
		_status.DriverTime = GetTickCount();
		if (_status.OutgoingPackets < _maxPackets)
		{
			int offset = _outgoingOffset + (_status.OutgoingPackets * _packetSize);
			for (int i = 0; i < _packetSize; i++)
				_commAccessor[offset + i] = packet[i];
			_status.OutgoingPackets++;
			*((CommStatus*)&_commAccessor[_statusOffset]) = _status;
		}
		else
		{
			//reset buffer
			OutputDebugString(L"Buffer full\n");
			_status.OutgoingPackets = 0;
			*((CommStatus*)&_commAccessor[_statusOffset]) = _status;
		}
		ReleaseMutex(_commAccessLock);
	}
}

char *CShMem::ReadIncomingPackets(int *count)
{
	*count = 0;
	char *result = nullptr;
	if (WaitForSingleObject(_commAccessLock, 100) == WAIT_OBJECT_0)
	{
		_status = *((CommStatus*)&_commAccessor[_statusOffset]);
		_status.DriverTime = GetTickCount();
		if (_status.IncomingPackets > 0)
		{
			result = new char[_status.IncomingPackets * _packetSize];
			int pos = 0;
			for (auto p = 0; p < _status.IncomingPackets; p++)
			{				
				auto currOffset = _incomingOffset + (p * _packetSize);
				for (auto i = 0; i < _packetSize; i++)
					result[pos++] = _commAccessor[currOffset + i];
			}
			*count = _status.IncomingPackets;
			_status.IncomingPackets = 0;			
		}
		*((CommStatus*)&_commAccessor[_statusOffset]) = _status;
		ReleaseMutex(_commAccessLock);
	}
	return result;
}

void CShMem::WriteScreen(int eye, char *screenData, int size)
{
	auto si = (ScreenInfo*)_screenAccessor[eye];	
	auto dd = _screenAccessor[eye] + sizeof(ScreenInfo);
	if (WaitForSingleObject(_screenAccessLock[eye], 100) == WAIT_OBJECT_0)
	{
		si->Size = size;		
		memcpy(dd, screenData, size);
		ReleaseMutex(_screenAccessLock[eye]);
	}
}